src/beamlime/constructors/providers.py

```diff
@@ -199,12 +203,54 @@ def __hash__(self) -> int:
     def __repr__(self) -> str:
         return f"Provider({self.call_name}, *{self.args}, **{self.keywords})."
 
+    @classmethod
+    def __copy__(cls, _obj: Provider[Product]) -> Provider[Product]:
+        return cls(_obj)
+
+    def __deepcopy__(self, _: dict[Any, Any]) -> Provider[Product]:
+        from copy import copy
+
+        return copy(self)
+
+
+Constructor = Union[Provider[Product], partial[Product], Callable[..., Product]]
+
+
+def _dummy_call(*args: Any, **kwargs: Any) -> None:
+    ...
+
+
+class CachedProviderCalledWithDifferentArgs(Exception):
+    ...
+
+
+class CachedProvider(Provider[Product]):
+    def __init__(
+        self, _constructor: Constructor[Product], /, *args: Any, **kwargs: Any
+    ) -> None:
+        from functools import lru_cache
+
+        super().__init__(_constructor, *args, **kwargs)
+        self.cached_constructor = lru_cache(maxsize=1)(self._constructor)
```

Comment:
This creates reference loops because `self` refers to the cache and the cache refers to `self` via the bound first argument of `self._constructor` .

I am not aware of a good solution here. You may have to implement caching yourself. Since you already check the arguments in the call, this shouldn't be too difficult.

Reply:
I tried to implement caching myself but hashing `args` and `kwargs` is complicated when you don't know what they will be so I thought it'll be easier to just use `lru_cache`.
And... I don't see the reference loop though, `lru_cache` wraps the `self._constructor` and the wrapper will have the cache.